src/CuratedModuleGate.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-FileCopyrightText: 2025 Lido <info@lido.fi>
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.24;
+
+import { AccessControlEnumerableUpgradeable } from "@openzeppelin/contracts-upgradeable/access/extensions/AccessControlEnumerableUpgradeable.sol";
+import { MerkleProof } from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+
+import { AssetRecoverer } from "./abstract/AssetRecoverer.sol";
+import { PausableUntil } from "./lib/utils/PausableUntil.sol";
+
+import { ICuratedModule } from "./interfaces/ICuratedModule.sol";
+import { IMerkleGate } from "./interfaces/IMerkleGate.sol";
+import { ICuratedModuleGate } from "./interfaces/ICuratedModuleGate.sol";
+import { NodeOperatorManagementProperties } from "./interfaces/ICSModule.sol";
+import { IOperatorsData } from "./interfaces/IOperatorsData.sol";
+import { ICSAccounting } from "./interfaces/ICSAccounting.sol";
+
+/// @notice Merkle gate for Curated Module v2 to create Node Operators
+///         and persist operator metadata (name, description) in a separate storage contract.
+contract CuratedModuleGate is
+    ICuratedModuleGate,
+    AccessControlEnumerableUpgradeable,
+    PausableUntil,
+    AssetRecoverer
+{
+    bytes32 public constant PAUSE_ROLE = keccak256("PAUSE_ROLE");
+    bytes32 public constant RESUME_ROLE = keccak256("RESUME_ROLE");
+    bytes32 public constant RECOVERER_ROLE = keccak256("RECOVERER_ROLE");
+    bytes32 public constant SET_TREE_ROLE = keccak256("SET_TREE_ROLE");
+
+    /// @inheritdoc ICuratedModuleGate
+    ICuratedModule public immutable MODULE;
+
+    /// @inheritdoc ICuratedModuleGate
+    ICSAccounting public immutable ACCOUNTING;
+
+    /// @inheritdoc ICuratedModuleGate
+    IOperatorsData public immutable OPERATORS_DATA;
+
+    /// @inheritdoc IMerkleGate
+    bytes32 public treeRoot;
+
+    /// @inheritdoc IMerkleGate
+    string public treeCid;
+
+    /// @inheritdoc ICuratedModuleGate
+    uint256 public curveId;
+
+    /// @dev Tracks whether an address already consumed its eligibility
+    mapping(address => bool) internal _consumedAddresses;
+
+    constructor(address module, address operatorsData) {
+        if (module == address(0)) revert ZeroModuleAddress();
+        if (operatorsData == address(0)) revert ZeroOperatorsDataAddress();
+        MODULE = ICuratedModule(module);
+        ACCOUNTING = MODULE.accounting();
+        OPERATORS_DATA = IOperatorsData(operatorsData);
+        _disableInitializers();
+    }
+
+    function initialize(
+        uint256 _curveId,
+        bytes32 _treeRoot,
+        string calldata _treeCid,
+        address admin
+    ) external initializer {
+        __AccessControlEnumerable_init();
+        if (admin == address(0)) revert ZeroAdminAddress();
+        if (_curveId == ACCOUNTING.DEFAULT_BOND_CURVE_ID())
+            revert InvalidCurveId();
```

Comment:
I think this condition is irrelevant for CM v2 since there will be no default permissionless gate